### ISSUE TITLE
KeyError in describe_location on user movement

### ISSUE THREAD

**Issue by zhammer** [brandon-rhodes/python-adventure#21]
Hey Brandon--

I hate to open an issue on code that hasn't been touched for two years, but I was playing ```python-adventure``` today (I recently watched your python/unix keynote and was so happy you ported adventure to python) and hit a bug.

I was going to try to do some debugging myself but I saw a few too many spoilers in the source code.

The exception didn't crash the game but it did occasionally reoccur. I tried rerunning python-adventure with the same sequence of commands a few times but didn't see the error again. I believe the error has to do with the random dwarf movements.

I've attached the markdown snippet of the first error traceback. I also have a gist of the [complete game session](https://gist.github.com/zhammer/0ba3114392005c3d9ce760b6b7673fd4) as well as a filtered version with just [my python commands](https://gist.github.com/zhammer/54bbcea326bccf748aa5a4402af7cd6e).

This obviously isn't very important, but I thought you may want to give it a look!
```py
 >>> down
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/personal/python-adventure/adventure/prompt.py", line 18, in __repr__
    output = self.game.do_command(self.words)
  File "~/personal/python-adventure/adventure/game.py", line 437, in do_command
    self._do_command(words)
  File "~/personal/python-adventure/adventure/game.py", line 506, in _do_command
    self.dispatch_command(words)
  File "~/personal/python-adventure/adventure/game.py", line 553, in dispatch_command
    return self.do_motion(word1)
  File "~/personal/python-adventure/adventure/game.py", line 701, in do_motion
    self.move_to(move.action)
  File "~/personal/python-adventure/adventure/game.py", line 186, in move_to
    self.move_dwarves()
  File "~/personal/python-adventure/adventure/game.py", line 325, in move_dwarves
    self.describe_location()
  File "~/personal/python-adventure/adventure/game.py", line 379, in describe_location
    self.write(obj.messages[prop])
KeyError: 0
```

**Comment by brandon-rhodes** [comment]
Thank you! I had of course hoped after all these years that the code was finally perfect, but if it's not, then it's best that I know so that I can tackle it at some point. Let's leave this issue open until either I have time to return to the project, or so that other folks maybe learning Python see that there's an interesting bug to tackle.

**Comment by zhammer** [comment]
Sounds good. I'll try giving this a look once I've gotten a bit further in the game.

**Comment by zhammer** [comment]
I found the bug. There is a confusion in the code between the ```game.dwarf``` and ```game.dwarves``` objects that causes buggy behavior with dwarf interactions.

Here's a way-too-detailed bug explanation, mainly so that I can remember everything I looked at today.

### The original ```KeyError```

#### When does it happen?

You see a dwarf. You pick him up. You drop him. You perform another action. ```KeyError: 0```; the game crashes.

```
> w
THERE IS A THREATENING LITTLE DWARF IN THE ROOM WITH YOU!

YOU'RE IN DIRTY PASSAGE.

THERE IS A LITTLE AXE HERE.

> take dwarf
OK

> drop dwarf
OK

> d
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/zhammer/code/python-adventure/adventure/__main__.py", line 48, in <module>
    loop()
  File "/Users/zhammer/code/python-adventure/adventure/__main__.py", line 45, in loop
    baudout(game.do_command(words))
  File "/Users/zhammer/code/python-adventure/adventure/game.py", line 437, in do_command
    self._do_command(words)
  File "/Users/zhammer/code/python-adventure/adventure/game.py", line 506, in _do_command
    self.dispatch_command(words)
  File "/Users/zhammer/code/python-adventure/adventure/game.py", line 553, in dispatch_command
    return self.do_motion(word1)
  File "/Users/zhammer/code/python-adventure/adventure/game.py", line 769, in do_motion
    self.move_to()
  File "/Users/zhammer/code/python-adventure/adventure/game.py", line 186, in move_to
    self.move_dwarves()
  File "/Users/zhammer/code/python-adventure/adventure/game.py", line 325, in move_dwarves
    self.describe_location()
  File "/Users/zhammer/code/python-adventure/adventure/game.py", line 379, in describe_location
    self.write(obj.messages[prop])
KeyError: 0
```

#### What causes the bug?

##### You see a dwarf.
There's a dwarf in ```game.dwarves``` where ```dwarf.room == game.loc```.
```py
>>> w.game.loc
<room 108 at 0x10177bda0>
>>> pprint([dwarf for dwarf in w.game.dwarves if dwarf.room == w.game.loc][0].__dict__)
{'has_seen_adventurer': True,
 'old_room': <room 108 at 0x10177bda0>,
 'room': <room 108 at 0x10177bda0>}
```


##### You pick him up.

When the user enters ```take(dwarf)```, ```def dispatch_command(self, words)``` is called with ```words``` passed as the tuple: ```('take', 'dwarf')```. 

```dispatch_command``` finds the ```model.Word``` mapped to each user-input string using the ```game.vocabulary``` dict.

```py
# game.py:519-521 with added debug prints
print('string words: {}'.format(words)) # DEBUG
words = [ self.vocabulary.get(word) for word in words ]
if None in words:
   return self.dont_understand()
print('mapped Words: {}'.format(words)) # DEBUG

# cli
>>> take(dwarf)
string words: ('take', 'dwarf')
mapped Words: [<Word take>, <Word dwarf>]
OK
>>> pprint(w.game.vocabulary['dwarf'].__dict__)
{'kind': 'noun',
 'n': 1017,
 'synonyms': [<Word dwarf>, <Word dwarves>],
 'text': 'dwarf'}
```

```dispatch_command``` then finds the ```Object``` to which ```<Word dwarf>``` is mapped via a call to ```self.referent```.

```py
# game.py:574
obj = self.referent(noun)
print('obj: {} found from noun: {}'.format(obj, noun)) # DEBUG

# cli
>>> take(dwarf)
obj: <Object 17 dwarf/dwarves 105f102e8> found from noun: "<Word dwarf>"
```

The issue here is that ```<Object 17 dwarf/dwarves 105f102e8>``` is different from the ```Dwarf``` the user has encountered.

```<Object 17 dwarf/dwarves 105f102e8>``` is a ```model.Object``` in ```game.object_list``` that is parsed from section 4 of ```adventure.dat```on startup. This is the object referenced by ```game.dwarf```.
```py
>>> w.game.vocabulary['dwarf']
<Word dwarf>
>>> w.game.referent(_)
<Object 17 dwarf/dwarves 10d9c10b8>
>>> w.game.dwarf
<Object 17 dwarf/dwarves 10d9c10b8>
>>> w.game.object_list[16]
<Object 17 dwarf/dwarves 10d9c10b8>
>>> pprint(w.game.dwarf.__dict__)
{'contents': None,
 'inventory_message': '',
 'is_fixed': False,
 'is_toting': False,
 'is_treasure': False,
 'messages': {},
 'n': 17,
 'names': ['dwarf', 'dwarves'],
 'prop': 0,
 'rooms': [],
 'starting_rooms': []}
```
This is a different object from the ```model.Dwarf``` in the user's room, which is a member of ```game.dwarves```.
```py
>>> pprint(w.game.dwarves)
[<adventure.model.Dwarf object at 0x10d9e2f60>,
 <adventure.model.Dwarf object at 0x10d9e2f98>,
 <adventure.model.Dwarf object at 0x10d9e2fd0>,
 <adventure.model.Dwarf object at 0x10d9e8048>]
>>> [dwarf for dwarf in w.game.dwarves if dwarf.room == w.game.loc][0]
<adventure.model.Dwarf object at 0x10d9e2fd0>
>>> _.__dict__
{'room': <room 108 at 0x10d95eb70>, 'old_room': <room 106 at 0x10d95ea90>, 'has_seen_adventurer': True}
```

After some logic (```game.py:574-582```) correctly checks whether a dwarf in ```game.dwarves``` is in the same room as the user, ```dispatch_command``` dispatches a call to ```t_carry```, passing the ```game.dwarf``` (```<Object 17 dwarf/dwarves 10d9c10b8>```) sentinel dwarf object as the ```obj``` argument.

```t_carry``` calls ```obj.carry``` on ```game.dwarf``` which switches ```is_toting``` to ```True```.

```py
>>> w.game.dwarf.is_toting
False
>>> w.game.inventory
[<Object 2 lamp/headlamp/lantern 10d9bcc50>, <Object 4 cage 10d9bccf8>]
>>> take(dwarf)
OK

>>> w.game.dwarf.is_toting
True
>>> w.game.inventory
[<Object 2 lamp/headlamp/lantern 10d9bcc50>, <Object 4 cage 10d9bccf8>, <Object 17 dwarf/dwarves 10d9c10b8>]
```

##### You drop him.

Calling ```drop(dwarf)``` triggers the same sequence of commands in ```dispatch_command``` is with ```take(dwarf)```. ```t_drop``` is called, with its ```obj``` argument being a reference to ```game.dwarf```.

This calls ```obj.drop``` on ```game.dwarf```, which sets ```is_toting``` to ```False``` AND also sets ```game.dwarf```s room to ```game.loc```.

```py
# model.py:138-140
def drop(self, room):
    self.rooms[:] = [ room ]
    self.is_toting = False

#cli
>>> w.game.dwarf.is_toting, w.game.dwarf.rooms
(False, [])
>>> take(dwarf)
OK

>>> w.game.dwarf.is_toting, w.game.dwarf.rooms
(True, [])
>>> drop(dwarf)
OK

>>> w.game.dwarf.is_toting, w.game.dwarf.rooms
(False, [<room 108 at 0x10ae2e0f0>])
```

##### You perform another action.

When a user performs a new action, ```describe_location``` is called. Part of ```describe_location``` writes the associated message of each object in```game.objects_here```, which is the list of objects that are in the user's room.

```py
for obj in self.objects_here:
   ...
   self.write(obj.messages[prop])
```

Since ```game.dwarf``` is a sentinel ```Object``` representing 'dwarf' and isn't intended to ever actually be 'in' a room, ```game.dwarf``` does not have any ```messages```.

```py
# 'spelunker', a tangible object, has a populated message dict
>>> w.game.object_list[15]
<Object 16 magazine/issue/spelunker/"spel 10ae89588>
>>> _.messages
{0.0: 'THERE ARE A FEW RECENT ISSUES OF "SPELUNKER TODAY" MAGAZINE HERE.\n'}

# 'dwarf', a 'sentinel' object, has an empty message dict
>>> w.game.object_list[16]
<Object 17 dwarf/dwarves 10ae895f8>
>>> _.messages
{}
```

##### ```KeyError: 0```; the game crashes.

Indexing into the empty ```game.dwarf.messages``` dictionary throws a ```KeyError```. ```game.dwarf``` should have never been in ```game.objects_here``` at all.

##### Expected behavior

I don't have access to the original Fortran copy, but I tried the version on http://www.amc.com/shows/halt-and-catch-fire/exclusives/colossal-cave-adventure. This is the behavior on that game, which makes sense to me.

```
There is no way to go that direction.
There is a threatening little dwarf in the room with you!

> take(dwarf)

I don't understand that!
```

#### Bugginess on ```attack```

```attack``` also has some buggy behavior, though it doesn't cause a crash. The ```attack(dwarf)``` command will call ```t_attack``` on the ```game.dwarf``` sentinel object. The ```attack``` command (without the 'dwarf' noun) will call ```t_attack``` on the actual ```model.Dwarf``` in the user's room.

I added a print statement to the top of ```t_attack``` to print out the ```obj``` arg being passed in the following examples.
```py
def t_attack(self, verb, obj):  #9124  (but control goes to 9120 first)
    print('t_attack obj: {}'.format(obj)) # DEBUG
```

##### ```attack(dwarf)```

What happens here is the same as with ```take(dwarf)``` or ```drop(dwarf)```. ```dispatch_command``` eventually maps the ```'dwarf'``` noun to ```game.dwarf``` and passes ```game.dwarf``` to ```t_attack```.

```py
>>> attack(dwarf)
t_attack obj: <Object 17 dwarf/dwarves 106dc82e8>
WITH WHAT?  YOUR BARE HANDS?
```

```t_attack``` handles the case where ```obj``` is ```self.dwarf```, printing the 'WITH WHAT?' message.

```py
# game.py:1336-1339
elif obj is self.dwarf:
    if self.is_closed:
        self.wake_repository_dwarves()
        return
    self.write_message(49)
```

##### ```attack``` without a noun

```py
>>> attack
t_attack obj: <adventure.model.Dwarf object at 0x106df0240>
THERE IS NOTHING HERE TO ATTACK.
```

Without a noun, ```i_attack``` is called. ```i_attack``` checks for ```dangers```: any ```enemies``` that are in the user's room. This list includes the ```model.Dwarf```s from ```game.dwarves```. If there is just one danger, ```i_attack``` calls ```t_attack``` on the danger (in this case the dwarf in the room).
```py
# game.py
def i_attack(self, verb):  #9120
    enemies = [ self.snake, self.dragon, self.troll, self.bear ]
    if self.dwarf_stage >= 2:
        enemies.extend(self.dwarves)
    dangers = list(filter(self.is_here, enemies))
    print('enemies: {}\ndangers: {}'.format(enemies, dangers)) # DEBUG
    if len(dangers) > 1:
        return self.ask_verb_what(verb)
    if len(dangers) == 1:
        return self.t_attack(verb, dangers[0])

# cli
>>> attack
enemies: [<Object 11 snake 10ebafeb8>, <Object 31 dragon 10ebb4400>, <Object 33 troll 10ebb44e0>, <Object 35 bear 10ebb4588>, <adventure.model.Dwarf object at 0x10ebcef60>, <adventure.model.Dwarf object at 0x10ebcef98>, <adventure.model.Dwarf object at 0x10ebcefd0>, <adventure.model.Dwarf object at 0x10ebd9048>]
dangers: [<adventure.model.Dwarf object at 0x10ebcefd0>]
t_attack obj: <adventure.model.Dwarf object at 0x10ebcefd0>
```

The ```t_attack``` conditional only checks if ```obj is self.dwarf```. Since ```obj``` here is a ```model.Dwarf``` in ```game.dwarves```, ```t_attack``` defaults to calling ```sef.write_message(44)```, which is the 'THERE IS NOTHING HERE TO ATTACK' message.

##### Expected behavior

Also from playing with the AMC website version:

```
> attack

With what? Your bare hands?
There is a threatening little dwarf in the room with you!
One sharp, nasty knife is thrown at you!
It misses!

> attack dwarf

With what? Your bare hands?
There is a threatening little dwarf in the room with you!
One sharp, nasty knife is thrown at you!
It misses!
```

### Fix

Fixing this just requires a few code changes to correctly handle ```game.dwarf``` vs ```game.dwarves```. ```game.dwarf``` seems to be the sentinel dwarf object used for understanding player commands. ```game.dwarves``` represent the actual ```model.Dwarf```s the player encounters.

I've forked the repo; once I have some time I'll try to work on a patch for this! Really love the game, and this is like a great easter egg.

**Comment by brandon-rhodes** [comment]
It _never_ would have occurred to me to try picking up a dwarf — thanks for uncovering this, and thanks for writing up your investigation in such detail! I look forward to seeing your pull request, if you find the time to get one written!

**Comment by zhammer** [comment]
^ accident

**Comment by brandon-rhodes** [comment]
(@zhammer I'm re-assigning this to myself since it looks like you didn't find the time for a pull request — let me know if I missed it!)

**Comment by zhammer** [comment]
ah yes i've gotten distracted with other work. if you do get around to doing this feel free to tag me on the PR! if you don't get around to it i'll remind myself to put up a PR soon.

**Comment by zhammer** [comment]
@brandon-rhodes I put up a PR! it is a simple 'catch this situation' fix, as i'm a little nervous to make too many changes to the code.

Added a test case to make sure that when we try to pick up a dwarf, we get a `dont_understand` error message!

**Comment by brandon-rhodes** [comment]
I merged your PR (thanks!), and then made two improvements.

First, not needing more than one dwarf in the room, I simplified the test to do
```
game.dwarves[0].room = game.rooms[75]
```

And second, I tracked down the root problem. No objects were getting marked as `is_fixed` because of a bug in the import routine! So the check:
```
        if obj.is_fixed or len(obj.rooms) > 1:
```
was never triggering because objects were marked as being impossible to pick up.

Let me know if the issue now looks resolved from your end, and thanks again for letting me know about it!
